cscs-checks/microbenchmarks/gpu/pointer_chase/src/linked_list.hpp

```diff
@@ -0,0 +1,383 @@
+#include "Xdevice/smi.hpp"
+
+/*
+ *
+ * Singly linked list implementation for GPUs
+ *
+ */
+
+
+__global__ void clockLatency(int * clk)
+{
+  // This returns the clock latency when reading the 64-bit clock counter.
+  clk[0] = XClockLatency<int>();
+}
+
+
+void printClockLatency(char * nid, int dev)
+{
+  /* Prints the latency of reading the clock cycles */
+  int * clk_d;
+  int clk;
+  XSetDevice(dev);
+  XMalloc((void**)&clk_d, sizeof(int));
+  clockLatency<<<1,1>>>(clk_d);
+  XDeviceSynchronize();
+  XMemcpy(&clk, clk_d, sizeof(int), XMemcpyDeviceToHost);
+  XFree(clk_d);
+  printf("[%s] The clock latency on device %d is %d cycles.\n", nid, dev, clk);
+}
+
+
+/*
+ * Linked list definitions
+ */
+
+struct Node
+{
+  /* The node */
+  Node * next = nullptr;
+  char _padding[8*NODE_PADDING];
+};
+
+
+/*
+ *  Kernels and device functions
+ */
+
+__global__ void initialize_list(Node * buffer, int stride = 1)
+{
+  /* List serial initializer.
+   * - buffer: where the list is to be placed.
+   * - stride: argument controls the number of empty node spaces
+   *   in between two consecutive list nodes.
+   */
+
+  // Set the head
+  Node * prev = new (&(buffer[0])) Node();
+
+  // Init the rest of the list
+  for (int n = 1; n < NODES; n++)
+  {
+    Node * temp = new (&(buffer[n*stride])) Node();
+    prev->next = temp;
+    prev = temp;
+  }
+
+}
+
+__global__ void initialize_random_list(Node * buffer, uint32_t *indices)
+{
+  /* List random initializer
+   * - buffer: where the list is to be placed.
+   * - indices: array containing the node ordering indices as offsets in
+   *     the buffer.
+   */
+
+  // Set the head
+  Node * prev = new (&(buffer[indices[0]])) Node();
+
+  // Init the rest of the list
+  for (int n = 1; n < NODES; n++)
+  {
+    Node * temp = new (&(buffer[indices[n]])) Node();
+    prev->next = temp;
+    prev = temp;
+  }
```

Comment:
Close the list with `prev->next = buffer + indices[0];`

Reply:
Same here.